### ISSUE TITLE
None gradients for codebook embeddings

### ISSUE THREAD

**Issue by kristinaivanova97** [EdoardoBotta/RQ-VAE-Recommender#24]
Hello!
First of all thank you for you work, it really helpful.
Sorry if I misunderstood something, but while analyzing gradients (as I have a suspicion that rqvae do not learn) I found out that codebook embeddings have None gradients, Is it supposed to be? In article it is written that codebook should have gradients while learning.

Piece of code before optimizer step:

for name, weight in model.named_parameters(): 
    print(name, weight.grad)
    
optimizer.step()

The output:
layers.0.embedding.weight None
layers.1.embedding.weight None
layers.2.embedding.weight None

While other parameters have gradients

Sorry if my English is not so good, hope you will understand and looking forward to your reply

**Comment by karapostK** [comment]
Hey!

I checked as well for STE and yes, you're right the gradient of the embeddings are None. Also looking at the parameters weights, the embeddings do not change across iterations.

I'm going to check why but, at least from a theoretical pov, the embeddings should receive an update from the rqvae_loss in the backward pass.

(P.S. not the author, just checking the code base)

**Comment by karapostK** [comment]
Okay, now it seems that it's working. 
The issues seems to happen when passing `embs` to `rqvae_loss = self.rqvae_loss(residuals, embs)`
As you noticed, detaching the embedding from the graph causes that no update arrives to the Embeddings in the Quantize class.

I'll post a solution soon

**Comment by EdoardoBotta** [comment]
thanks a lot for the contributions to both, this is very helpful. I agree on the issue with the gradients not flowing properly through the rq_vae loss due to the detach calls. 

It seems to me that https://github.com/EdoardoBotta/RQ-VAE-Recommender/pull/26 solves the issue for STE but it is most likely wrong for ROTATION_TRICK because the current implementation of ROTATION_TRICK is using soft embeddings computed by `app_emb = weights @ codebook`, while `org_emb` is computing hard embeddings using `org_emb = self.get_item_embeddings(ids)`.

I was thinking of actually removing these detach calls altogher from happening within the Quantize module and handling this piece logic in the RQVAE forward directly. This would allow the detach to be applied on the embeddings that go through the decoder and not be applied to the embedding.

I hacked a sketch of this approach in https://github.com/EdoardoBotta/RQ-VAE-Recommender/tree/fix_codebook_grad and will try to put together a cleaner implementation over the weekend.

**Comment by karapostK** [comment]
From my understanding of gradient flow:

- STE:
Reconstruction Loss updates the Encoder and Decoder. Vocabulary Loss updates the Codebooks
- Gumbel Softmax:
Both losses update all parameters.
- Rotation Trick:
Similar to STE.

Would using `detach` in the RQVAE work across all these approaches?

**Comment by EdoardoBotta** [comment]
Yeah, one would need to check and have a conditional based on QuantizeForwardMode within RQVAE forward as well. I agree with you that Karpathy's approach looks cleaner.

**Comment by EdoardoBotta** [comment]
Pushed the fix: https://github.com/EdoardoBotta/RQ-VAE-Recommender/commit/f2b505db13f2c73a8527403c1880b1bb600765e9. Tested that both ROTATION_TRICK and STE gradients are not None and fully trained the RQVAE for 200,000 iterations reaching much lower loss than before (ROTATION_TRICK: 0.08 vs. 0.17, STE: 0.1 vs. was exploding before).

**Comment by kristinaivanova97** [comment]
Seems like now code works good, thank you very much for updates and comments!
